compass/turbo_compass.py

```diff
@@ -0,0 +1,365 @@
+    r"""
+    Modifies args list in place, setting 'arg_name' argument to 'val'.
+    If 'arg_name' does not exist already, it is created.
+    """
+    if not isinstance(arg_name, str) or not isinstance(val, str):
+        raise ValueError(f"{arg_name} and {val} should be strings!")  # pragma: no cover
+    for i in range(len(args) - 1):
+        if args[i] == arg_name:
+            args[i + 1] = str(val)
+            return
+    args.append(arg_name)
+    args.append(val)
+
+
+def pop_argument(args: List[str], arg: str) -> Tuple[List[str], str]:
+    r"""
+    Removes the argument from the parameter list and returns the new parameter
+    list and the popped parameter's value. If the argument is not present, returns
+    the original arguments and the empty string.
+    e.g. pop_argument(['compass', '--turbo', '0.95']) == (['compass'], '0.95')
+    after the operation.
+    """
+    if arg not in args:
+        return args[:], ""
+    for i in range(len(args) - 1):
+        if args[i] == arg:
+            return args[:i] + args[(i + 2):], args[i + 1]
+    assert(False)
+
+
+def turbo_compass_entry() -> None:
+    logger.info("******** Turbo Compass started ********")
+    # Extract turbo arguments.
+    compass_parsed_args = compass.main.parseArgs()
+    requested_cv_spearman_r2, increments, min_pct_meet_sr2_requirement, max_iters =\
+        compass_parsed_args['turbo'],\
+        compass_parsed_args['turbo_increments'],\
+        compass_parsed_args['turbo_min_pct'],\
+        compass_parsed_args['turbo_max_iters']
+    logger.info(
+        f"Running Turbo Compass with arguments:\n"
+        f"--turbo={requested_cv_spearman_r2}\n"
+        f"--turbo-increments={increments}\n"
+        f"--turbo-min-pct={min_pct_meet_sr2_requirement}\n"
+        f"--turbo-max-iters={max_iters}")
+
+    # Remove turbo arguments from the command line call, and create the CompassOracle for the new command line.
+    compass_args = sys.argv[:]
+    compass_args, _ = pop_argument(compass_args, '--turbo')
+    compass_args, _ = pop_argument(compass_args, '--turbo-increments')
+    compass_args, _ = pop_argument(compass_args, '--turbo-min-pct')
+    compass_args, _ = pop_argument(compass_args, '--turbo-max-iters')
+    logger.info("Turbo Compass creating the CompassOracle ...")
+    compass_matrix_oracle = CompassOracle(compass_args[:])
+
+    model =\
+        IterativeMCMWithGuaranteedSpearmanR2(
+            cv_model=lambda state:
+                TrainValSplitCVMatrixCompletionModel(
+                    ExcludeConstantColumnsModelWrapper(
+                        ColumnNormalizerModelWrapper(
+                            MatrixCompletionFastALS(
+                                n_factors=max(1, int(min(state.R, state.C) * state.sampled_density * 0.5)),
+                                lam=10.0,
+                                n_epochs=100,
+                                verbose=False))),
+                    train_ratio=0.8,
```

Comment:
One simple improvement after discussing with Richard Zhang is to make the size of the cross-validation set adaptive:
- When we start off we have little data, so maybe a 80/20 split produces too small a validation set size.
- Once we have more data, a 80/20 split might be too big, i.e. we are being too conservative as the algorithm progresses (because we shrink the training set size more than necessary), and as a consequence calling Compass more than we should to meet the user's specified reconstruction quality.

Reply:
It sounds to me that what would be ideal would be to use a fixed CV size for each reaction, say size 50. In that case I would need to implement another concretion for `CVMatrixCompletionModel` which performs CV splitting based on a fixed CV size, rather than on a percentage.

Reply:
(or, rather than creating a new concretion, cleverly factor out the CV-splitting behavior to get the new behavior via composition)

Reply:
Or maybe, it's easier to just add a `max_cv_size` argument to the `TrainValSplitCVMatrixCompletionModel` API.